### ISSUE TITLE
Set the specific input type.

### ISSUE THREAD

**Issue by adriannelee** [cds-snc/notification-planning#478]
# Fields

Visibility: 3
Evaluation: Partially Supports
Success Criteria: 1.3.5: Identify Input Purpose
Needs design: 

### Description
When an input takes a specific type of value (e.g. e-mail, phone number, URL, etc.) use the related input type. This makes it possible for the browser to perform their own client side validation and presenting input specific entry.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input\_types

![]()

## Related research
[It was difficult using dictation to remove a space in the 'Sender email address' field
](https://airtable.com/appWwAmHwDLtpIyko/tbl38n7ofWYBuezFc/viwEUFgBd0KLqH7oo/reccdWxkWLVVmGAcd?blocks=hide)

# QA
## scenario 1 - invalid characters
- Navigate to `/add-service`
- Enter a **service name**
- Enter an invalid **sending email address name**: `Canadian Digital Service - Service Numerique Canadién`
- Click **Continue**
- [x] Ensure it gets fixed to: `canadian.digital.service-service.numerique.canadien`
- [x] Ensure you can continue to the next screen in the wizard

## scenario 2 - valid characters
- Navigate to `/add-service`
- Enter a **service name**
- Enter a valid **sending email address name**: `canadian.digital.service-service.numerique.canadien`
- Click **Continue**
- [x] Ensure you can continue to the next screen in the wizard without any warning

**Comment by andrewleith** [comment]
Did a quick comb through for places we may need to add a specific type to an input:
- [ ] `/sign-in` -> email
- [ ] `/register` -> email,mobile
- [ ] `/contact` -> email
- [ ] `/services/d6aa2c68-a2d9-4437-ab19-3ae8eb202553/api/safelist` -> email address, -> mobile
- [ ] `/services/8d6a24a2-d2a7-4b6b-9785-bc6805a65f53/service-settings/email_from` -> sending email address
- [ ] `/add-service?current_step=choose_email_from` -> sending email address
- [ ] `/user-profile/email` -> email address
- [ ] `/user-profile/mobile-number` -> mobile #
- [ ] `/forgot-password` -> email

If you know of any I've missed, please let me know.

**Comment by yaelberger-commits** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/478?planning-poker) with ZenHub @andrewleith

**Comment by yaelberger-commits** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/478?planning-poker) with ZenHub @jimleroyer @jzbahrai @iokpala @sastels

**Comment by adriannelee** [comment]
@andrewleith Maybe it's just on the `Create a new service` page where the email input type is missing? That's where the dictation software input spaces into the field. (I'm not totally sure if the input type would fix this issue - let me know!)

**Comment by adriannelee** [comment]
@andrewleith Hm maybe the input type needs to be set to "email" for `/services/8d6a24a2-d2a7-4b6b-9785-bc6805a65f53/service-settings/email_from`? Not sure what the difference between "email" and "sending email address is" :/ ?

Not sure if this is already the case, maybe the inputmode needs to be added too to solve the dictation issue? (from the mozilla link above)

> inputmode
> Global value valid for all elements, it provides a hint to browsers as to the type of virtual keyboard configuration to use when editing this element or its contents. Values include none, text, tel, url, email, numeric, decimal, and search.

**Comment by adriannelee** [comment]
@andrewleith Ohhh I see. 

> Block the typing of spaces? Remove them when they hit Continue?

If we can block the typing of spaces would work well here!

**Comment by amazingphilippe** [comment]
I think we can also look at the way the label and hint are written on this page. We could also try to find better a better example (currently passport-passeport)

**Comment by yaelberger-commits** [comment]
Andrew doing some mobile testing on phone, correct keyboards coming up

**Comment by amazingphilippe** [comment]
Andrew and I have been doing some thinking around this and we found a rabbit hole. In terms of interaction, blocking characters as the user is typing isn't a good idea, it creates frustration. This opened up a few ideas:
- We can remove or replace the spaces on validation
- We can instruct people to not use spaces, and give a validation error
- We can do both of the above

This could be fine for now, but it also opens up a few more UX opportunities:
- Could we generate the sender email from the service name?
- If so, could we improve how the service name is created bilingually?
- And other things that could improve the service creation experience
- Also improve how bilingual services are created and managed down the road. 

I think we need to draw the line somewhere for this sprint. Looking for other input from the team.

**Comment by adriannelee** [comment]
Ohh, hm. The issue that came up during testing was that for someone using dictation software to complete this field, they are experiencing frustration moving their cursor to the right space to delete the space. (e.g. they spoke "publicannouncement" aloud and the software typed in the space in between the words") So they knew to remove the spaces before hitting continue. But removing the spaces on validation is a great idea nonetheless.

I think generating the sender email from the service name would be interesting to explore. Maybe automatically suggests you something by taking in the English name and French name (if both exists) and allowing you to edit it if needed

**Comment by yaelberger-commits** [comment]
@adriannelee I think those are ideas worth exploring. Autogenerating the sending email address from the service name by default could help nudge Senders to create more consistency and better Service names as well. 

I would put that work into [Content Guidance](https://app.zenhub.com/workspaces/notify-planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/610) epic, as opposed to adding it on to this card. But I do agree it's worth exploring in another sprint in the future.

**Comment by yaelberger-commits** [comment]
@amazingphilippe I think where you drew the line above is right. The first 2/3 bullets seem like the first steps. The other bullets are great ideas to explore in a future Usability surge/spike. We can discuss this more at Backlog refinement today. Please feel free to raise it if I don't

**Comment by amazingphilippe** [comment]
Also considering mobile, even if we know most of our users aren't mobile users:
![image.png](https://images.zenhubusercontent.com/614b439e980ccd4137400d3e/3148d004-9e5e-4bca-a537-93de8d62341a)
